lib/authlogic/session/password.rb

```diff
@@ -150,16 +155,45 @@ def credentials
           end
         end
 
-        # Accepts the login_field / password_field credentials combination in
-        # hash form.
+        # Accepts the login_field / password_field credentials. `value` can be
+        # any of the following:
+        #
+        # ```
+        # { :login => "my login", :password => "my password", :remember_me => true }
+        # [my_user_object, true]
+        # [{ :login => "my login", :password => "my password", :remember_me => true }, :my_id]
+        # [my_user_object, true, :my_id]
+        # [{ priority_record: my_object }, :my_id]
+        # ```
+        #
+        # See `Foundation::InstanceMethods#credentials=` for further explanation.
+        #
+        # It's common to provide an ActiveSupport::Parameters instead of a Hash.
+        # We don't officially support that, but as long as you `permit` the
+        # expected keys (login_field, password_field) it's OK.
         def credentials=(value)
           super
-          values = parse_param_val(value) # add strong parameters check
+          potential_hash = value.is_a?(Array) ? value.first : value
+
+          # nil is allowed only for backwards compatibility. It's not really
+          # useful to do, e.g. `UserSession.new` with no arguments in practice.
+          return if potential_hash.nil?
 
-          if values.first.is_a?(Hash)
-            values.first.with_indifferent_access.slice(login_field, password_field).each do |field, val|
-              next if val.blank?
-              send("#{field}=", val)
+          if potential_hash.respond_to?(:to_h)
+            hash = potential_hash.to_h.with_indifferent_access
```

Comment:
The key here is that `AC::Params` behaves like a `HashWithIndifferentAccess` already so we only need to call `with_indifferent_access` on actual hashes.